File: engine/cpp/matching.cpp
// Price-time-priority matching: walk the opposite side from the best price inwards; within a level
// consume the FIFO from the head; every execution happens at the maker's (resting) price.
#include "order_book.hpp"

namespace lob {

i64 OrderBook::execute(int side, bool has_limit, i64 limit, i64 qty, i64 taker_id, i64 taker_owner, i64 ts) {
    SideMap& opp = side == BUY ? asks_ : bids_;
    while (qty > 0) {
        auto it = opp.begin();
        if (it == opp.end()) break;
        PriceLevel& lvl = it->second;
        if (has_limit) {
            if (side == BUY && lvl.price > limit) break;
            if (side == SELL && lvl.price < limit) break;
        }
        while (qty > 0 && lvl.head != nullptr) {
            Order* maker = lvl.head;
            const i64 fill = qty < maker->qty ? qty : maker->qty;
            const i64 remaining = maker->qty - fill;
            ++counters_.trades;
            counters_.traded_qty += fill;
            emit(EventType::TRADE, ts, [&](Event& e) {
                e.order_id = taker_id; e.side = side; e.price = maker->price; e.qty = fill; e.owner = taker_owner;
                e.maker_id = maker->id; e.maker_owner = maker->owner; e.maker_remaining = remaining;
            });
            qty -= fill;
            if (remaining == 0) {
                lvl.remove(maker);
                index_.erase(maker->id);
                pool_.release(maker);
            } else {
                lvl.reduce(maker, fill);
            }
        }
        if (lvl.count == 0) opp.erase(it);
    }
    return qty;
}

}  // namespace lob

File: engine/python/matching.py
"""Price-time-priority matching.

The matcher walks the opposite side from the best price inwards. Within a level it consumes
the FIFO queue from the head. Each execution happens at the *maker's* (resting) price.
Cost is O(k) for k consumed orders plus O(1) per fully consumed level.
"""
from __future__ import annotations

from typing import Callable, Optional

from engine.common import Side
from engine.python.price_levels import BookSide

# emit_trade(maker_node, executed_qty, maker_remaining_after)
TradeEmit = Callable[["object", int, int], None]


def match(
    opposite: BookSide,
    taker_side: Side,
    limit_price: Optional[int],
    qty: int,
    index: dict,
    emit_trade: TradeEmit,
) -> int:
    """Execute up to ``qty`` against ``opposite``; return the unfilled remainder.

    ``limit_price=None`` means a market order (no price bound). Filled makers are removed from
    ``index`` and emptied levels are removed from ``opposite``; partially filled makers keep
    their queue position.
    """
    while qty > 0:
        level = opposite.best_level()
        if level is None:
            break
        if limit_price is not None:
            if taker_side is Side.BUY and level.price > limit_price:
                break
            if taker_side is Side.SELL and level.price < limit_price:
                break
        while qty > 0 and level.head is not None:
            maker = level.head
            fill = qty if qty < maker.qty else maker.qty
            remaining = maker.qty - fill
            emit_trade(maker, fill, remaining)  # emitted before mutation so maker.qty is pre-fill
            qty -= fill
            if remaining == 0:
                level.remove(maker)
                del index[maker.order_id]
            else:
                level.reduce(maker, fill)
        if level.count == 0:
            opposite.remove_level(level.price)
    return qty
